File: packages/typia/src/programmers/misc/MiscIsCloneProgrammer.ts
import ts from "typescript";

import { IdentifierFactory } from "../../factories/IdentifierFactory";
import { StatementFactory } from "../../factories/StatementFactory";
import { TypeFactory } from "../../factories/TypeFactory";

import { IProgrammerProps } from "../../transformers/IProgrammerProps";
import { ITypiaContext } from "../../transformers/ITypiaContext";

import { FeatureProgrammer } from "../FeatureProgrammer";
import { IsProgrammer } from "../IsProgrammer";
import { FunctionProgrammer } from "../helpers/FunctionProgrammer";
import { MiscCloneProgrammer } from "./MiscCloneProgrammer";

export namespace MiscIsCloneProgrammer {
  export const decompose = (props: {
    context: ITypiaContext;
    functor: FunctionProgrammer;
    type: ts.Type;
    name: string | undefined;
  }): FeatureProgrammer.IDecomposed => {
    const is: FeatureProgrammer.IDecomposed = IsProgrammer.decompose({
      ...props,
      config: {
        equals: false,
      },
    });
    const clone: FeatureProgrammer.IDecomposed = MiscCloneProgrammer.decompose({
      ...props,
      validated: true,
    });
    return {
      functions: {
        ...is.functions,
        ...clone.functions,
      },
      statements: [
        ...is.statements,
        ...clone.statements,
        StatementFactory.constant({
          name: "__is",
          value: is.arrow,
        }),
        StatementFactory.constant({
          name: "__clone",
          value: clone.arrow,
        }),
      ],
      arrow: ts.factory.createArrowFunction(
        undefined,
        undefined,
        [IdentifierFactory.parameter("input", TypeFactory.keyword("any"))],
        ts.factory.createUnionTypeNode([
          clone.arrow.type ?? TypeFactory.keyword("any"),
          ts.factory.createTypeReferenceNode("null"),
        ]),
        undefined,
        ts.factory.createBlock(
          [
            ts.factory.createIfStatement(
              ts.factory.createPrefixUnaryExpression(
                ts.SyntaxKind.ExclamationToken,
                ts.factory.createCallExpression(
                  ts.factory.createIdentifier("__is"),
                  undefined,
                  [ts.factory.createIdentifier("input")],
                ),
              ),
              ts.factory.createReturnStatement(ts.factory.createNull()),
            ),
            ts.factory.createReturnStatement(
              ts.factory.createCallExpression(
                ts.factory.createIdentifier("__clone"),
                undefined,
                [ts.factory.createIdentifier("input")],
              ),
            ),
          ],
          true,
        ),
      ),
    };
  };

  export const write = (props: IProgrammerProps): ts.CallExpression => {
    const functor: FunctionProgrammer = new FunctionProgrammer(
      props.modulo.getText(),
    );
    const result: FeatureProgrammer.IDecomposed = decompose({
      ...props,
      functor,
    });
    return FeatureProgrammer.writeDecomposed({
      modulo: props.modulo,
      functor,
      result,
    });
  };
}

File: packages/typia/native/core/programmers/misc/MiscIsCloneProgrammer.go
package misc

import (
  shimast "github.com/microsoft/typescript-go/shim/ast"
  shimchecker "github.com/microsoft/typescript-go/shim/checker"
  nativecontext "github.com/samchon/typia/packages/typia/native/core/context"
  nativefactories "github.com/samchon/typia/packages/typia/native/core/factories"
  nativeprogrammers "github.com/samchon/typia/packages/typia/native/core/programmers"
  nativehelpers "github.com/samchon/typia/packages/typia/native/core/programmers/helpers"
  nativeinternal "github.com/samchon/typia/packages/typia/native/core/programmers/internal"
)

type miscIsCloneProgrammerNamespace struct{}

var MiscIsCloneProgrammer = miscIsCloneProgrammerNamespace{}

type MiscIsCloneProgrammer_DecomposeProps struct {
  Context nativecontext.ITypiaContext
  Functor *nativehelpers.FunctionProgrammer
  Type    *shimchecker.Type
  Name    *string
}

var miscIsCloneProgrammer_factory = shimast.NewNodeFactory(shimast.NodeFactoryHooks{})

func (miscIsCloneProgrammerNamespace) Decompose(props MiscIsCloneProgrammer_DecomposeProps) nativeinternal.FeatureProgrammer_IDecomposed {
  is := nativeprogrammers.IsProgrammer.Decompose(nativeprogrammers.IsProgrammer_DecomposeProps{
    Context: props.Context,
    Functor: props.Functor,
    Config:  nativeprogrammers.IsProgrammer_IConfig{Equals: false},
    Type:    props.Type,
    Name:    props.Name,
  })
  clone := MiscCloneProgrammer.Decompose(MiscCloneProgrammer_DecomposeProps{
    Context:   props.Context,
    Functor:   props.Functor,
    Type:      props.Type,
    Name:      props.Name,
    Validated: true,
  })
  statements := append([]*shimast.Node{}, is.Statements...)
  statements = append(statements, clone.Statements...)
  statements = append(statements,
    nativefactories.StatementFactory.Constant(nativefactories.StatementFactory_ConstantProps{Name: "__is", Value: is.Arrow}),
    nativefactories.StatementFactory.Constant(nativefactories.StatementFactory_ConstantProps{Name: "__clone", Value: clone.Arrow}),
  )
  cloneType := clone.Arrow.AsArrowFunction().Type
  if cloneType == nil {
    cloneType = nativefactories.TypeFactory.Keyword("any")
  }
  return nativeinternal.FeatureProgrammer_IDecomposed{
    Functions:  miscProgrammer_merge_functions(is.Functions, clone.Functions),
    Statements: statements,
    Arrow: miscIsCloneProgrammer_factory.NewArrowFunction(
      nil,
      nil,
      miscIsCloneProgrammer_factory.NewNodeList([]*shimast.Node{
        nativefactories.IdentifierFactory.Parameter("input", nativefactories.TypeFactory.Keyword("any"), nil),
      }),
      miscIsCloneProgrammer_factory.NewUnionTypeNode(miscIsCloneProgrammer_factory.NewNodeList([]*shimast.Node{
        cloneType,
        miscIsCloneProgrammer_factory.NewTypeReferenceNode(miscIsCloneProgrammer_factory.NewIdentifier("null"), nil),
      })),
      nil,
      miscIsCloneProgrammer_factory.NewToken(shimast.KindEqualsGreaterThanToken),
      miscIsCloneProgrammer_factory.NewBlock(miscIsCloneProgrammer_factory.NewNodeList([]*shimast.Node{
        miscIsCloneProgrammer_factory.NewIfStatement(
          miscIsCloneProgrammer_factory.NewPrefixUnaryExpression(
            shimast.KindExclamationToken,
            miscIsCloneProgrammer_factory.NewCallExpression(
              miscIsCloneProgrammer_factory.NewIdentifier("__is"),
              nil,
              nil,
              miscIsCloneProgrammer_factory.NewNodeList([]*shimast.Node{miscIsCloneProgrammer_factory.NewIdentifier("input")}),
              shimast.NodeFlagsNone,
            ),
          ),
          miscIsCloneProgrammer_factory.NewReturnStatement(miscIsCloneProgrammer_factory.NewKeywordExpression(shimast.KindNullKeyword)),
          nil,
        ),
        miscIsCloneProgrammer_factory.NewReturnStatement(miscIsCloneProgrammer_factory.NewCallExpression(
          miscIsCloneProgrammer_factory.NewIdentifier("__clone"),
          nil,
          nil,
          miscIsCloneProgrammer_factory.NewNodeList([]*shimast.Node{miscIsCloneProgrammer_factory.NewIdentifier("input")}),
          shimast.NodeFlagsNone,
        )),
      }), true),
    ),
  }
}

func (miscIsCloneProgrammerNamespace) Write(props nativecontext.IProgrammerProps) *shimast.Node {
  functor := nativehelpers.NewFunctionProgrammer(miscCloneProgrammer_method_text(props.Modulo))
  result := MiscIsCloneProgrammer.Decompose(MiscIsCloneProgrammer_DecomposeProps{
    Context: props.Context,
    Functor: functor,
    Type:    props.Type,
    Name:    props.Name,
  })
  return nativeinternal.FeatureProgrammer.WriteDecomposed(nativeinternal.FeatureProgrammer_WriteDecomposedProps{
    Modulo:  props.Modulo,
    Functor: functor,
    Result:  result,
  })
}

func miscProgrammer_merge_functions(groups ...map[string]*shimast.Node) map[string]*shimast.Node {
  output := map[string]*shimast.Node{}
  for _, group := range groups {
    for key, value := range group {
      output[key] = value
    }
  }
  return output
}
